### PR TITLE
Deserialize `BitSet`

### PR DESCRIPTION
**Pull request by emmaling27**
We need the ability to read a mutable `BitSet` from disk so we can update when documents in previous segments are deleted.

### CODE REVIEW

File: common/src/bitset.rs [hunk 1]
```diff
@@ -196,6 +196,30 @@ impl BitSet {
...
+        let mut tinysets_buf = Vec::new();
+        reader.read_to_end(&mut tinysets_buf)?;
+        let mut tinysets = Vec::new();
+        let mut len = 0;
+        for chunk in tinysets_buf.chunks_exact(8) {
+            let slice: &[u8; 8] = chunk.try_into().unwrap();
+            let tinyset = TinySet::deserialize(slice.clone());
+            len += tinyset.len() as u64;
+            tinysets.push(tinyset);
+        }
```
Comment (sujayakar) [inline common/src/bitset.rs]: you could avoid a few allocations by (1) presizing the buffers and (2) parsing the tinysets as we read the file:
```rust
let buckets = num_buckets(max_value);
let mut tinysets = Vec::with_capacity(buckets);

for _ in 0..buckets {
    let mut buf = [0; 8];
    reader.read_exact(&mut buf)?;
    let tinyset = TinySet::deserialize(buf);
    tinysets.push(tinyset);
}

File: common/src/bitset.rs [hunk 2]
```diff
@@ -196,6 +196,30 @@ impl BitSet {
         Ok(())
     }
 
+    pub fn deserialize(mut reader: impl Read) -> io::Result<Self> {
+        let mut max_buf = vec![0; 4];
```
Comment (sujayakar) [inline common/src/bitset.rs]: i have a different suggestion below, but you can avoid a heap allocation by using an array on the stack for these small buffers:
```rust
let mut max_buf = [0; 4];
reader.read_exact(&mut max_buf)?;
```

File: common/src/bitset.rs [hunk 3]
```diff
@@ -196,6 +196,30 @@ impl BitSet {
...
+        reader.read_exact(&mut max_buf)?;
+        let max_value =
+            u32::from_le_bytes(max_buf.try_into().expect("Max value should be 4 bytes"));
```
Comment (sujayakar) [inline common/src/bitset.rs]: tantivy already uses the awesome byteorder crate, which lets you read numerics out of a stream with a given endianness directly!
```rust
use byteorder::{LittleEndian, ReadBytesExt};

let max_value = reader.read_u32::<LittleEndian>()?;
```

**Review by sujayakar — APPROVED** [review]
awesome!! some minor rust suggestions inline but lgtm overall.

**Review by sjudd — APPROVED** [review]

File: common/src/bitset.rs (hunk 1 above)
Comment (emmaling27) [reply common/src/bitset.rs]: sweet! thanks for these optimization suggestions.

**Pull request merged**